crates/router/src/core/payment_methods/cards.rs

```
@@ -216,6 +225,216 @@ pub async fn get_or_insert_payment_method(
+                        nick_name: card.nick_name,
+                        card_network: None,
+                        card_isin: None,
+                        card_issuer: None,
+                        card_type: None,
+                        saved_to_locker: true,
+                    });
+
+                    let updated_pmd = updated_card.as_ref().map(|card| {
+                        PaymentMethodsData::Card(CardDetailsPaymentMethod::from(card.clone()))
+                    });
+                    let pm_data_encrypted =
+                        create_encrypted_payment_method_data(key_store, updated_pmd).await;
+
+                    let pm_update = storage::PaymentMethodUpdate::AdditionalDataUpdate {
+                        payment_method_data: pm_data_encrypted,
+                        status: Some(enums::PaymentMethodStatus::Active),
+                        locker_id: Some(locker_id),
+                        payment_method: req.payment_method,
+                    };
+
+                    db.update_payment_method(payment_method, pm_update)
+                        .await
+                        .change_context(errors::ApiErrorResponse::InternalServerError)
+                        .attach_printable("Failed to add payment method in db")?;
+
+                    return Ok(services::ApplicationResponse::Json(pm_resp));
+                }
+            } else {
+                let pm_update = storage::PaymentMethodUpdate::StatusUpdate {
+                    status: Some(enums::PaymentMethodStatus::Inactive),
+                };
+
+                db.update_payment_method(payment_method, pm_update)
+                    .await
+                    .change_context(errors::ApiErrorResponse::InternalServerError)
+                    .attach_printable("Failed to udpate payment method in db")?;
+
+                return Err(report!(errors::ApiErrorResponse::InternalServerError)
+                    .attach_printable("Failed to add card to locker"));
```

aren't u loosing/ignoring the error stack report?

---

Could you describe a bit on how this should be refactored?

---

instead of using `it let` on `resp`, use match and catch the `Err(err)` and change_context from this `err` to `ApiErrorResponse`. Something like below

`Err(err) => err.change_context(ApiErrorResponse)?`

---

Done. Could you check?